text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+To demonstrate this case, consider the following service and component:
+
+```js
+const Config = Service.extend({
+  polling: {
+    shouldPoll: false,
+    pollInterval: -1,
+  },
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  },
+});
+```
+```js
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    let { shouldPoll, pollInterval } = this.config.polling;
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+Let's walk through the flow here:
+
+1. The `SomeComponent` component is rendered for the first time, instantiating
+   the `Config` service (assuming this the first time it has ever been
+   accessed). The service's init hook kicks off an async request to get the
+   configuration from a remote URl.
+2. The tracked `pollInterval` property first accesses the service injection,
+   which is a computed property. The property is detected and added to the
+   tracked stack.
+3. We then access the plain, undecorated `polling` object. Because it is
+   is not tracked and not a computed property, tracked does not know that it
+   could update in the future.
+4. Sometime later, the async request returns with the configuration object. We
+   set it on the service, but because our tracked getter did not know this
+   property would update, it does not invalidate.
+
+In order to prevent this from happening, user's will have to use `get` when
+accessing any values which may be set with `set`, and are not computed
+properties.
+
+```js
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    let shouldPoll = get(this, 'config.polling.shouldPoll');
+    let pollInterval = get(this, 'config.polling.pollInterval');
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+The reverse, however, is not true - computed properties will be able to add
+tracked properties, and listen to dependencies explicitly. In some cases, this
+may be preferable, though tracked getter should be the conventional standard
+with the long term goal of removing all explicit dependencies.
+
+## How we teach this
+
+There are three different aspects of tracked properties which need to be
+considered for the learning story:
+
+1. **General usage.** Which properties should I mark as tracked? How do I
+   consume them? How do I trigger changes?
+2. **Interop with classic systems.** How do I safely consume tracked properties
+   from classic classes and computeds? How do I safely consume classic APIs from
+   tracked properties?
+3. **Interop with non-Ember systems.** How do I tell my app that something has
+   changed in MobX objects, RxJS objects, Redux, etc.
+
+### General Usage
+
+The mental model with tracked properties is that anything _mutable_ should be
+tracked. If a value will ever change, it should have the `@tracked` decorator
+attached to it.
```

Comment:
Some subset of internal component concerns, although mutable, probably don't need to be tracked (and do not require the use of `set()` today)

Example: I kick off a `setInterval` immediately after my component is first rendered, and need to `clearInterval` as the component is torn down. The return value of `setInterval` should be kept around to facilitate this.

This state is never rendered, never used as an element of some derived state, and thus never needs change tracking. This statement

> anything _mutable_ should be tracked

Should be qualified appropriately